### PR TITLE
[TySan] A Type Sanitizer (Clang)

### PR DESCRIPTION
**Pull request by fhahn**
This patch introduces the runtime components of a type sanitizer: a sanitizer for type-based aliasing violations.

C/C++ have type-based aliasing rules, and LLVM's optimizer can exploit these given TBAA metadata added by Clang. Roughly, a pointer of given type cannot be used to access an object of a different type (with, of course, certain exceptions). Unfortunately, there's a lot of code in the wild that violates these rules (e.g. for type punning), and such code often must be built with -fno-strict-aliasing. Performance is often sacrificed as a result. Part of the problem is the difficulty of finding TBAA violations. Hopefully, this sanitizer will help.

The Clang changes seems mostly formulaic, the one specific change being that when the TBAA sanitizer is enabled, TBAA is always generated, even at -O0.

Clang's TBAA representation currently has a problem representing unions, as demonstrated by the one XFAIL'd test in the runtime patch. We'll update the TBAA representation to fix this, and at the same time, update the sanitizer.

Based on https://reviews.llvm.org/D32199.

https://github.com/llvm/llvm-project/pull/76259 (LLVM support)

### CODE REVIEW

**Review by erichkeane — COMMENTED** [review]
The clang changes are ok, but this needs some level of documentation/release notes, which I don't see in the clang release.  As this is a part of a larger feature, do we intend to push that later?

Also, the clang-format suggestion makes sense.

**Review by erichkeane — COMMENTED** [review]
A pair of minor changes requested, else this looks about right?  Not sure who the right person to approve this is though

File: clang/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -1027,6 +1027,10 @@ Sanitizers
   <https://clang.llvm.org/docs/SanitizerSpecialCaseList.html>`_. See that link
   for examples.
 
+- Introduced an experimental Type Sanitizer, activated by using the
+  -fsanitize=type flag. This sanitizer detects violations of C/C++ type-based
```
Comment (AaronBallman) [inline clang/docs/ReleaseNotes.rst:1031]:
```suggestion
  ``-fsanitize=type`` flag. This sanitizer detects violations of C/C++ type-based
```

File: clang/include/clang/Basic/Features.def [hunk 2]
```diff
@@ -102,6 +102,7 @@ FEATURE(numerical_stability_sanitizer, LangOpts.Sanitize.has(SanitizerKind::Nume
 FEATURE(memory_sanitizer,
         LangOpts.Sanitize.hasOneOf(SanitizerKind::Memory |
                                    SanitizerKind::KernelMemory))
+FEATURE(type_sanitizer, LangOpts.Sanitize.has(SanitizerKind::Type))
```
Comment (AaronBallman) [inline clang/include/clang/Basic/Features.def:105]: None of these sanitizers are features. :-( I think this is the correct thing to do for consistency, but at the same time, we keep adding more sanitizers and we keep making this problem worse.

Nothing to change here, just me grumbling. :-D

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 3]
```diff
@@ -6630,7 +6631,8 @@ CodeGenModule::GetAddrOfConstantStringFromLiteral(const StringLiteral *S,
   if (Entry)
     *Entry = GV;
 
-  SanitizerMD->reportGlobal(GV, S->getStrTokenLoc(0), "<string literal>");
+  SanitizerMD->reportGlobalToASan(GV, S->getStrTokenLoc(0), "<string literal>");
+  // FIXME: Should we also report to the TySan?
```
[the remark this answers is not in the document]
Comment (fhahn) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Yep, merged the code to report globals again, thanks!

File: clang/lib/CodeGen/CodeGenModule.cpp [hunk 4]
```diff
@@ -5740,7 +5740,8 @@ void CodeGenModule::EmitGlobalVarDefinition(const VarDecl *D,
   if (NeedsGlobalCtor || NeedsGlobalDtor)
     EmitCXXGlobalVarDeclInitFunc(D, GV, NeedsGlobalCtor);
 
-  SanitizerMD->reportGlobal(GV, *D, NeedsGlobalCtor);
+  SanitizerMD->reportGlobalToASan(GV, *D, NeedsGlobalCtor);
```
[the remark this answers is not in the document]
Comment (fhahn) [reply clang/lib/CodeGen/CodeGenModule.cpp]: Actually on current `main`, `reportGlobal` reports not only to Asan, but also memsan and hwsan. Updated to include reporting globals to TySan, thanks!

**Review by fhahn — COMMENTED** [review]
Thanks, as this depends on LLVM & compiler-rt patches, so it should only be merged once those other PRs are also approved.

Motivation for the sanitizer is enabling `-fpointer-tbaa` by default (https://github.com/llvm/llvm-project/pull/117244)

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 5]
```diff
@@ -1441,8 +1441,10 @@ collectSanitizerRuntimes(const ToolChain &TC, const ArgList &Args,
     if (SanArgs.needsScudoRt()) {
       SharedRuntimes.push_back("scudo_standalone");
     }
-    if (SanArgs.needsTsanRt())
+    if (SanArgs.needsTsanRt() && SanArgs.linkRuntimes())
       SharedRuntimes.push_back("tsan");
+    if (SanArgs.needsTysanRt())
+      StaticRuntimes.push_back("tysan");
```
Comment (efriedma-quic) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:1447]: Did you mean `SharedRuntimes.push_back`?

File: clang/lib/Driver/ToolChains/CommonArgs.cpp [hunk 6]
```diff
@@ -1441,8 +1441,10 @@ collectSanitizerRuntimes(const ToolChain &TC, const ArgList &Args,
     if (SanArgs.needsScudoRt()) {
       SharedRuntimes.push_back("scudo_standalone");
     }
-    if (SanArgs.needsTsanRt())
+    if (SanArgs.needsTsanRt() && SanArgs.linkRuntimes())
```
Comment (efriedma-quic) [inline clang/lib/Driver/ToolChains/CommonArgs.cpp:1444]: How is this change to tsan related to tysan?

**Review by efriedma-quic — APPROVED** [review]
LGTM with comments addressed.  (I haven't looked carefully at the non-clang parts; this is just approval for the clang changes.)

File: clang/lib/Driver/ToolChains/CommonArgs.cpp (hunk 6 above)
Comment (fhahn) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp]: Nope, I think this was left over from a merge, removed, thanks!

File: clang/lib/Driver/ToolChains/CommonArgs.cpp (hunk 5 above)
Comment (fhahn) [reply clang/lib/Driver/ToolChains/CommonArgs.cpp]: Fixed, thanks!

**Review by fhahn — COMMENTED** [review]
> LGTM with comments addressed. (I haven't looked carefully at the non-clang parts; this is just approval for the clang changes.)


Thanks! Any chance you would be able to help with the review of the LLVM side as well? https://github.com/llvm/llvm-project/pull/76259

**Comment by fhahn** [comment]
All 3 parts have been approved and the LLVM part just landed. I am planning on landing the Clang part and the compiler-rt part later today

File: clang/lib/CodeGen/SanitizerMetadata.cpp [hunk 7]
```diff
@@ -105,11 +106,32 @@ void SanitizerMetadata::reportGlobal(llvm::GlobalVariable *GV,
                                            GV, Loc, Ty, "init");
 
   GV->setSanitizerMetadata(Meta);
+
+  if (Ty.isNull() || !CGM.getLangOpts().Sanitize.has(SanitizerKind::Type) ||
```
Comment (fhahn) [inline clang/lib/CodeGen/SanitizerMetadata.cpp:110]: We also need to exclude globals defined externally as well I think: https://github.com/llvm/llvm-project/pull/120565

**Comment by cor3ntin** [comment]
Sorry for the very late review but do we have clang documentation for this?

**Comment by AaronBallman** [comment]
> Sorry for the very late review but do we have clang documentation for this?

+1, we do need documentation for this before Clang 20 ships. CC @fhahn

**Pull request merged**